### PR TITLE
ISPN-5816 Event Logger

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.jboss.org/browse/ISPN-5816

Implements a per-node event logger which stores events in a cache with 7-day retention. 
Also incorporates https://issues.jboss.org/browse/ISPN-6035 (i.e. allow internal caches to be queryable)

### CODE REVIEW

**Comment by tristantarrant** [comment]
This PR is currently missing the ability to collect the local events using distexec, since the DefaultExecutorService requires a clustered cache, and I'm using a local one for resilience (since we need to track events like splits/merges).

File: query/src/main/java/org/infinispan/query/impl/LifecycleManager.java [hunk 1]
```diff
@@ -93,8 +94,9 @@
    @Override
    public void cacheStarting(ComponentRegistry cr, Configuration cfg, String cacheName) {
       InternalCacheRegistry icr = cr.getGlobalComponentRegistry().getComponent(InternalCacheRegistry.class);
-      if (!icr.isInternalCache(cacheName)) {
-         cr.registerComponent(new ReflectionMatcher(null), ReflectionMatcher.class);
+      if (!icr.isInternalCache(cacheName) || icr.internalCacheHasFlag(cacheName, Flag.QUERYABLE)) {
```
Comment (anistor) [inline query/src/main/java/org/infinispan/query/impl/LifecycleManager.java]: I think the ```internalCacheHasFlag(cacheName, Flag.QUERYABLE)``` check has to be done in the LifecycleManager of remote query too, for completeness, even if that is not used in this PR.

Comment (tristantarrant) [inline query/src/main/java/org/infinispan/query/impl/LifecycleManager.java]: Thanks for pointing that out

**Comment by tristantarrant** [comment]
This is now ready for review. The "distributed" part has been deferred to https://issues.jboss.org/browse/ISPN-6078

File: core/src/main/java/org/infinispan/util/logging/events/EventLogCategory.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package org.infinispan.util.logging.events;
+
+/**
+ * EventLogCategory.
+ *
+ * @author Tristan Tarrant
+ * @since 8.2
+ */
+public enum EventLogCategory {
+   LIFECYCLE,
```
Comment (vblagoje) [inline core/src/main/java/org/infinispan/util/logging/events/EventLogCategory.java]: @tristantarrant why not leave some guidelines when each category is supposed to be used?

File: server/integration/event-logger/src/main/java/org/infinispan/server/eventlogger/ServerEventLogger.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+   public void log(EventLogLevel level, EventLogCategory category, String message) {
+      textLog(level, category, message);
+      eventLog(new ServerEventImpl(level, category, timeService.instant(), message));
+   }
+
+   void textLog(EventLogLevel level, EventLogCategory category, String message) {
+      LogFactory.getLogger(category.toString()).log(level.toLoggerLevel(), message);
+   }
+
+   void eventLog(ServerEventImpl event) {
```
Comment (vblagoje) [inline server/integration/event-logger/src/main/java/org/infinispan/server/eventlogger/ServerEventLogger.java]: Maybe not even expose this method? ServerEventImpl thus stays encapsulated in ServerEventLogger

Comment (tristantarrant) [inline server/integration/event-logger/src/main/java/org/infinispan/server/eventlogger/ServerEventLogger.java]: you mean, hoist the code within eventLog to log ?

Comment (vblagoje) [inline server/integration/event-logger/src/main/java/org/infinispan/server/eventlogger/ServerEventLogger.java]: Nevermind @tristantarrant I overlooked that it is package protected. When I wrote the comment I thought it was public

**Comment by vblagoje** [comment]
Integrated with 7fdf87e 7635bbb 2823e60 Thanks @tristantarrant

**Pull request closed without merging**